File: usecases/src/main/java/com/zeyad/usecases/network/RestApi.java
package com.zeyad.usecases.network;

import android.support.annotation.NonNull;

import java.util.List;
import java.util.Map;

import io.reactivex.Flowable;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.PATCH;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.PartMap;
import retrofit2.http.Streaming;
import retrofit2.http.Url;

/**
 * RestApi for retrieving data from the network.
 */
public interface RestApi {

    @NonNull
    @GET
    Flowable<Object> dynamicGetObject(@Url String url);

    @NonNull
    @GET
    Flowable<Object> dynamicGetObject(@Url String url, boolean shouldCache);

    @NonNull
    @GET
    Flowable<List> dynamicGetList(@Url String url);

    @NonNull
    @GET
    Flowable<List> dynamicGetList(@Url String url, boolean shouldCache);

    @NonNull
    @POST
    Flowable<Object> dynamicPost(@Url String url, @Body RequestBody body);

    @NonNull
    @PUT
    Flowable<Object> dynamicPut(@Url String url, @Body RequestBody body);

    @NonNull
    @PATCH
    Flowable<Object> dynamicPatch(@Url String url, @Body RequestBody requestBody);

    @NonNull
    @DELETE
    Flowable<Object> dynamicDelete(@Url String url);

    @NonNull
    @Streaming
    @GET
    Flowable<ResponseBody> dynamicDownload(@Url String fileUrl);

    @NonNull
    @Multipart
    @POST
    Flowable<Object> dynamicUpload(@Url String url, @PartMap() Map<String, RequestBody> partMap,
            @Part List<MultipartBody.Part> file);
}

File: usecases/src/main/java/com/zeyad/usecases/network/RestApi.kt
package com.zeyad.usecases.network

import io.reactivex.Flowable
import io.reactivex.Single
import okhttp3.MultipartBody
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.http.*

/**
 * RestApi for retrieving data from the network.
 */
interface RestApi {

    @GET
    fun dynamicGetObject(@Url url: String): Flowable<Any>

    @GET
    fun dynamicGetObject(@Url url: String, shouldCache: Boolean): Flowable<Any>

    @GET
    fun dynamicGetList(@Url url: String): Flowable<List<Any>>

    @GET
    fun dynamicGetList(@Url url: String, shouldCache: Boolean): Flowable<List<Any>>

    @POST
    fun dynamicPost(@Url url: String, @Body body: RequestBody): Single<Any>

    @PUT
    fun dynamicPut(@Url url: String, @Body body: RequestBody): Single<Any>

    @PATCH
    fun dynamicPatch(@Url url: String, @Body requestBody: RequestBody): Single<Any>

    @DELETE
    fun dynamicDelete(@Url url: String, @Body requestBody: RequestBody): Single<Any>

    @Streaming
    @GET
    fun dynamicDownload(@Url fileUrl: String): Single<ResponseBody>

    @Multipart
    @POST
    fun dynamicUpload(@Url url: String, @PartMap partMap: Map<String, RequestBody>,
                      @Part file: List<MultipartBody.Part>): Single<Any>
}
